### ISSUE TITLE
DbContext.SaveChangesAsync() Method Doesn't Adhere to CommandTimeout

### ISSUE THREAD

**Issue by MattWood21** [dotnet/efcore#8846]
When using the DbContext.SaveChangesAsync() method, the resulting task does not adhere to the DbContext.Database.CommandTimeout value.

### Steps to reproduce
I have a very simple DbContext isntance which is created by Dependency Injection:

Startup.cs:
```c#
            services.AddDbContext<apsdbContext>(options =>
            {
                options.UseNpgsql(_config["APSDB_CONNECTION"]);                
            });
```
And a repository that uses the context, also in Startup.cs:
```c#
            services.AddScoped<IAPSDBRepository, APSDBRepository>();
```
The repository receives the context via the constructor and adds results to it like this:
```c#
        private apsdbContext _context;

        public APSDBRepository(apsdbContext context)
        {
            _context = context;
        }

        public async Task<int> AddResult(Result result)
        {
            _context.Database.SetCommandTimeout(5);            
            _context.Add(result);
            return await _context.SaveChangesAsync();            
        }
```
After manually locking the Result table, this command never times out and deadlocks until the table lock is release.

If I change the AddResult method to the following, it times out after five seconds as expected:
```c#
        public int AddResult(Result result)
        {
            _context.Database.SetCommandTimeout(5);            
            _context.Add(result);
            return _context.SaveChanges();            
        }
```

### Further technical details
EF Core version: 1.1.2
Database Provider: Npgsql.EntityFrameworkCore.PostgreSQL 1.1.0
Operating system: Windows 10 (Development), Windows Server 2012 R2 (Production)
IDE: VS2017 Pro

**Comment by AndriySvyryd** [comment]
@Padagi Async operations need to be waited upon. Try the flowing
```C#
public int AddResult(Result result)
{
    _context.Database.SetCommandTimeout(5);            
    _context.Add(result);
    return _context.SaveChangesAsync().Result;            
}
```

Note that in production code `AddResult` should return `Task<int>` that would be awaited on in turn.

**Comment by Padagi** [comment]
@AndriySvyryd My apologies, I didn't think someone would look at this that quickly... I've updated the original issue with the EXACT code that we have in production today. It includes await and returns a Task<int>. I think I updated it while you were responding.


I just ran a quick test with SaveChangesAsync().Result, which runs the command synchronously anyway, but it still does not work. I tried the following and it also does not time out as it does if you run the SaveChanges() method:
```C#
        public int AddResult(Result result)
        {
            _context.Database.SetCommandTimeout(5);            
            _context.Add(result);
            return _context.SaveChangesAsync().Result;
        }
```

This code still hangs indefinitely until the underlying provider connection (in this case Npgsql) times out based on its own global settings.
